api/src/main/java/com/github/retrooper/packetevents/util/mappings/SynchronizedRegistriesHandler.java

```diff
@@ -80,12 +81,32 @@ private SynchronizedRegistriesHandler() {
     public static void handleRegistry(
             User user, ClientVersion version,
             ResourceLocation registryName,
-            List<RegistryElement> elements
+            List<RegistryElement> elements,
+            boolean caching
     ) {
         RegistryEntry<?> registry = REGISTRY_KEYS.get(registryName);
         if (registry != null) {
-            IRegistry<?> syncedRegistry = registry.createFromElements(elements, version);
-            user.putUserRegistry(syncedRegistry);
+            SimpleRegistry<?> syncedRegistry = registry.createFromElements(elements, version);
+            if (caching) {
+                // Cache the synchronized registries within each VersionedRegistry.
+                if (registry.baseRegistry instanceof VersionedRegistry) {
+                    VersionedRegistry<?> versionedRegistry = (VersionedRegistry<?>) registry.baseRegistry;
+
+                    //Did we already cache these? If so, skip...
+                    if (versionedRegistry.getSynchronizedRegistry() != null) return;
+
+                    versionedRegistry.synchronizeRegistry(syncedRegistry);
+                }
+
+                // Dimension type registry is the one exception that we always store for each user.
+                if (syncedRegistry.getRegistryKey().equals(DimensionTypes.getRegistry().getRegistryKey())) {
+                    user.putUserRegistry(syncedRegistry);
```

Comment:
If we are on proxy, we store all registries for each user.
If we are not on proxy, we only store the dimension type registry for each user, and cache everything else.